Brand new arays doc. A small amount of work remains to be done. (#4307)
yugabyte/yugabyte-db: docs/content/latest/api/ysql/datatypes/type_array/functions-operators/comparison.md

---

## Comparison operators overview

**Purpose:** Each of the comparison operators returns `true` or `false` according to the outcome of the particular comparison test between the input [LHS and RHS](https://en.wikipedia.org/wiki/Sides_of_an_equation) arrays.

**Signature**

These operators all have the same signature, thus:

```
input value:       anyarray, anyarray
return value:      boolean
```
**Notes:** These operators require that the LHS and RHS arrays have the same datatype. (It's the same rule for the comparison of scalars.) However, they do _not_ require that the arrays have identical geometric properties. Rules are defined so that a difference between one or more of these properties does not mean that they are necessarily simply unequal. Rather, the LHS array might be deemed to be less than, or greater than, the RHS array. We need, therefore, to understand the comparison algorithm.

### Comparison criteria

These are the unique characteristics of an array with respect to the algorithm that compares two array values:

- the actual values, compared pairwise in row-major order;
- the cardinality;
- the number of dimensions;
- the lower bound(s) on each dimension.

> The other geometric properties (the length(s) along each dimension and the upper bound(s) on each dimension) can be derived from the properties that the bullets list.

Reviewer: ```suggestion
The other geometric properties (the length(s) along each dimension and the upper bounds on each dimension) can be derived from the properties that the bullets list.
```

There is, of course, a well-defined priority among the comparisons. Briefly, value comparison is done first. Then, but only if no difference is detected, are the geometric properties compared.

### Pairwise comparison of values

> The first comparison test scans the values in each of the LHS and RHS arrays in row-major order (see [here](../functions-operators/properties/#joint-semantics)) and does a pairwise comparison. Notably, the comparison rule non-negotiably uses `is not distinct from` semantics. Moreover, when a `not null` array value is pairwise compared with a `null` value, the `not null` value is deemed to be _less than_ the `null` value.

Reviewer: ```suggestion
The first comparison test scans the values in each of the LHS and RHS arrays in row-major order (see [Joint semantics](../functions-operators/properties/#joint-semantics)) and does a pairwise comparison. Notably, the comparison rule non-negotiably uses `is not distinct from` semantics. Moreover, when a `not null` array value is pairwise compared with a `null` value, the `not null` value is deemed to be _less than_ the `null` value.
```

Notice that the two arrays might not have the same cardinality. If all possible pairwise comparisons result in equality, then the array with the greater cardinality is deemed to be greater than the other array, and the other geometric properties are not considered.

### The priority of differences among the geometric properties

The previous section stated the rule that the cardinality comparison has the highest priority among the geometric property comparisons. And that this rule kicks in only of all possible value comparisons result in equality.

When _both_ all possible value comparisons _and_ the cardinality comparison result in equality, then the comparison between the number of dimensions has a higher priority than the comparison between the lower bound(s). Of course, the array with the greater number of dimensions is deemed to be the greater array.

This means that the lower bound(s) are significant when two arrays are compared _only_ when they are identical in pairwise value comparison, cardinality, and the number of dimensions. Then the array with the greater lower bound, in dimension order, is deemed to be the greater array.

> The _"Equality and inequality semantics"_ section, [below](./#equality-and-inequality-semantics) demonstrates each of the rules that this _"Comparison operators overview"_ , section has stated.

Reviewer: ```suggestion
The [Equality and inequality semantics](./#equality-and-inequality-semantics) section demonstrates each of the rules that this _"Comparison operators overview"_  section has stated.
```

## Containment and overlap operators overview

These three operators are insensitive to the geometric properties of the two to-be-compared arrays.
- The two containment operators test if the distinct set of values in one array contains, or is contained by, the distinct set of values in the other array.
- The overlap operator tests if the distinct set of values in one array and the distinct set of values in the other array have at least one value in common.

> The _"Containment and overlap operators semantics"_ section, [below](./#containment-and-overlap-operators-semantics) demonstrates each of the rules that this section has stated.

Reviewer: ```suggestion
The [Containment and overlap operators semantics](./#containment-and-overlap-operators-semantics) section below demonstrates each of the rules that this section has stated.
```

------------------------------
 true
```

### The `>` and `>=` and `<=` and `<` and `<>` operators

These four operators implement the familiar inequality comparisons.
- The `>` operator returns `true` if the LHS array is greater than the RHS array.
- The `>=` operator returns `true` if the LHS array is greater than or equal to the RHS array.
- The `<=` operator returns `true` if the LHS array is less than or equal to the RHS array.
- The `<` operator returns `true` if the LHS array is less than the RHS array.

It's sufficient, therefore, to provide a simple example for just the `<` operator.
```postgresql
with
  v as (
    select
      (select array['a', 'b', 'c',  'd']::text[]) as a1,
      (select array['a', 'b', 'e',  'd']::text[]) as a2,
      (select      '{a,   b,  null,  d}'::text[]) as a3
  )
select
  (a1 < a2)::text as "'LESS THAN' comparison result 1",
  (a1 < a3)::text as "'LESS THAN' comparison result 2"
from v;
```
This is the result:
```
 'LESS THAN' comparison result 1 | 'LESS THAN' comparison result 2 
---------------------------------+---------------------------------
 true                            | true
```

### The `@>` and `<@` operators

> - The `@>` operator returns `true` if the LHS array contains the RHS array—i.e. if every distinct value in the RHS array is found among the LHS array's distinct values.

Reviewer: ```suggestion
- The `@>` operator returns `true` if the LHS array contains the RHS array—that is, if every distinct value in the RHS array is found among the LHS array's distinct values.
```

- The `<@` operator is the natural complement: it returns `true` if every distinct value in the LHS array is found among the RHS array's distinct values.

```postgresql
with
  v as (
    select
      (select array['a', 'b', 'c',  'd']::text[]) as a1,
      (select array['a',      'c'      ]::text[]) as a2
  )
select
  (a1 @> a2)::text as "CONTAINS comparison result",
  (a2 <@ a1)::text as "'IS CONTAINED BY' comparison result"
from v;
```
This is the result:
```
 CONTAINS comparison result | 'IS CONTAINED BY' comparison result 
----------------------------+-------------------------------------
 true                       | true
```

### The `&&` operator

> The `&&` operator returns `true` if the LHS and RHS arrays overlap—i.e. if they have at least one value in common. The definition of this operator makes it insensitive to which of the two to-be-compared is used on the LHS and which is used on the RHS.

Reviewer: ```suggestion
The `&&` operator returns `true` if the LHS and RHS arrays overlap—that is, if they have at least one value in common. The definition of this operator makes it insensitive to which of the two to-be-compared is used on the LHS and which is used on the RHS.
```

```postgresql
with
  v as (
    select
      (select array['a', 'b', 'c',  'd']::text[]) as a1,
      (select array['d', 'e', 'f',  'g']::text[]) as a2
  )
select
  (a1 && a2)::text as "'a1 OVERLAPS a2' comparison result",
  (a2 && a1)::text as "'a2 OVERLAPS a1' comparison result"
from v;
```
This is the result:
```
 'a1 OVERLAPS a2' comparison result | 'a2 OVERLAPS a1' comparison result 
------------------------------------+------------------------------------
 true                               | true
```

## Equality and inequality semantics

> This section demonstrates each of the rules that the _"Comparison operators overview"_ section, [above](./#comparison-operators-overview), stated.

Reviewer: ```suggestion
This section demonstrates each of the rules that the [Comparison operators overview](./#comparison-operators-overview) section above stated.
```

```postgresql

> -- Any two arrays can be compared without error if they have the same datatype.

Reviewer: ```suggestion
-- Any two arrays can be compared without error if they have the same data type.
```

  ------------------------------------------------------------------------------
  -- Illustrate "IS NOT DISTINCT FROM" semantics.
  declare
    v1 constant int := 1;
    v2 constant int := 1;
    n1 constant int := null;
    n2 constant int := null;
  begin
    assert
      (v1 = v2)                    and
      (v1 is not distinct from v2) and

      ((n1 = n2) is null)          and
      (n1 is not distinct from n2),
    'unexpected';
  end;

  ------------------------------------------------------------------------------
  -- Basic demonstration of equaliy when the geom. properties of
  -- the two arrays are identical.
  -- Shows that pairwise comparison uses ""IS NOT DISTINCT FROM" semantics and NOT
  -- the conventional NULL semantics used when scalars are compared.
  declare
    a constant int[] := '{10, null, 30}';
    b constant int[] := '{10, null, 30}'; -- Identical to a.
  begin
    assert
      (a = b),
    '"a = b" assert failed';

    -- Because of this, we don't need ever to write this.
    assert
      (a is not distinct from b),
    '"a is not distinct from b" assert failed';
  end;

  ------------------------------------------------------------------------------

>   -- Basic demonstration of inequaliy when the geom. properties of

Reviewer: ```suggestion
  -- Basic demonstration of inequality when the geometric properties of
```

Author: Of course code comments should ne fee of spelling errors. Do you have a spell checker that handles these sensibly without annoying false positives from the code itself?

But are abbreviations in comments really disallowed. Anything that readuces line length and the line count is highly desirable in code illustrations.

  -- the two arrays are identical.
  -- When the first difference is encountered in row-major order, the comparison
  -- is made. Other differences are irrelevant.
  declare
    a constant int[] := '{10, 20, 30}';
    b constant int[] := '{10, 19, 31}'; 
  begin
    assert
      (a <> b) and
      (a >  b) and
      (a >= b) and
      (b <= a) and
      (b <  a) ,
    '"a > b" assert failed';
  end;

  ------------------------------------------------------------------------------

>   -- Demonstration of inequaliy when the geom. properties of

Reviewer: ```suggestion
  -- Demonstration of inequality when the geometric properties of
```

  -- the two arrays are identical.

>   -- Here, the first pairwise difference is NOT NUL vs NULL.

Reviewer: ```suggestion
  -- Here, the first pairwise difference is NOT NULL versus NULL.
```
